src/renderer/atlas/common.h

```diff
@@ -0,0 +1,496 @@
+            return *this;
+        }
+
+        explicit operator bool() const noexcept
+        {
+            return _data != nullptr;
+        }
+
+        T& operator[](size_t index) noexcept
+        {
+            assert(index < _size);
+#pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+            return _data[index];
+        }
+
+        const T& operator[](size_t index) const noexcept
+        {
+            assert(index < _size);
+#pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+            return _data[index];
+        }
+
+        T* data() noexcept
+        {
+            return _data;
+        }
+
+        const T* data() const noexcept
+        {
+            return _data;
+        }
+
+        size_t size() const noexcept
+        {
+            return _size;
+        }
+
+        T* begin() noexcept
+        {
+            return _data;
+        }
+
+        const T* begin() const noexcept
+        {
+            return _data;
+        }
+
+        T* end() noexcept
+        {
+#pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+            return _data + _size;
+        }
+
+        const T* end() const noexcept
+        {
+#pragma warning(suppress : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+            return _data + _size;
+        }
+
+    private:
+        // These two functions don't need to use scoped objects or standard allocators,
+        // since this class is in fact an scoped allocator object itself.
+#pragma warning(push)
+#pragma warning(disable : 26402) // Return a scoped object instead of a heap-allocated if it has a move constructor (r.3).
+#pragma warning(disable : 26409) // Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
+        static T* allocate(size_t size)
+        {
+            if (!size)
+            {
+                return nullptr;
+            }
+            if constexpr (Alignment <= __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+            {
+                return static_cast<T*>(::operator new(size * sizeof(T)));
+            }
+            else
+            {
+                return static_cast<T*>(::operator new(size * sizeof(T), static_cast<std::align_val_t>(Alignment)));
+            }
+        }
+
+        static void deallocate(T* data) noexcept
+        {
+            if constexpr (Alignment <= __STDCPP_DEFAULT_NEW_ALIGNMENT__)
+            {
+                ::operator delete(data);
+            }
+            else
+            {
+                ::operator delete(data, static_cast<std::align_val_t>(Alignment));
+            }
+        }
+#pragma warning(pop)
+
+        void destroy() noexcept
+        {
+            std::destroy_n(_data, _size);
+            deallocate(_data);
+        }
+
+        T* _data = nullptr;
+        size_t _size = 0;
+    };
+
+    struct TargetSettings

```

Comment:
These structs, all the way down to `GenerationalSettings`, are new. They're all that's needed to represent the entire rendering state, and will hopefully replace the many getters/setters on `IRenderData` in the near term.